pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -684,28 +690,53 @@ public synchronized void clearWarnings() throws SQLException {
     firstWarning = null;
   }
 
-
+  @Override
   public void setReadOnly(boolean readOnly) throws SQLException {
     checkClosed();
+
+    if (readOnly == this.readOnly && cacheXactState) {
+      return; //Same read only state, not set it again.
+    }
+
     if (queryExecutor.getTransactionState() != TransactionState.IDLE) {
       throw new PSQLException(
           GT.tr("Cannot change transaction read-only property in the middle of a transaction."),
           PSQLState.ACTIVE_SQL_TRANSACTION);
     }
 
-    if (readOnly != this.readOnly) {
-      String readOnlySql
-             = "SET SESSION CHARACTERISTICS AS TRANSACTION " + (readOnly ? "READ ONLY" : "READ WRITE");
-      execSQLUpdate(readOnlySql); // nb: no BEGIN triggered.
-    }
+    String readOnlySql
+        = "SET SESSION CHARACTERISTICS AS TRANSACTION " + (readOnly ? "READ ONLY" : "READ WRITE");
+    execSQLUpdate(readOnlySql); // nb: no BEGIN triggered.
 
     this.readOnly = readOnly;
     LOGGER.log(Level.FINE, "  setReadOnly = {0}", readOnly);
   }
 
+  @Override
   public boolean isReadOnly() throws SQLException {
     checkClosed();
-    return readOnly;
+
+    if (!cacheXactState) {
+      ResultSet rs = null;
+      try {
+        rs = execSQLQuery("show transaction_read_only"); // nb: no BEGIN triggered
+        if (rs.next()) {
+          String transaction_read_only = rs.getString(1);
+          if ("on".equals(transaction_read_only)) {
+            this.readOnly = true;
```

Comment:
Isn't `this.readOnly = "on".equals(transaction_read_only);` enough?

Reply:
Sure, I will change that.